### DISCUSSION TITLE
Not recognizing L9 Attribute

### DISCUSSION THREAD

**Discussion by JeffreyDavidson** [larastan/larastan#1328] [General]
I'm trying to understand why I have anywhere that is trying to access currentTeam in my application it is coming back with the following error. `54     Access to an undefined property App\Models\User::$currentTeam.`

Is there a way to find out what properties Larastan has access to when running?

```php
namespace App\Models;
​
use App\Models\Concerns\CanJoinTeams;
use Illuminate\Database\Eloquent\Factories\HasFactory;
​
class User extends BaseModel
{
    use CanJoinTeams;
    use HasFactory;
}
```
​
```php
namespace App\Models\Concerns;
​
use App\Models\Team;
use Illuminate\Database\Eloquent\Casts\Attribute;
​
trait CanJoinTeams
{
    /**
     * Get the teams the user has been belonged to.
     *
     * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
     */
    public function teams()
    {
        return $this->belongsToMany(Team::class, 'team_user')
            ->withPivot(['joined_at', 'left_at']);
    }
​
    /**
     * Get the user's current team.
     *
     * @return \Illuminate\Database\Eloquent\Casts\Attribute
     */
    public function currentTeam(): Attribute
    {
        return Attribute::make(
            get: fn () => $this->teams()->wherePivotNull('left_at')->first(),
        );
    }
}
```
​
​
```php
namespace App\Models{
    /**
     * App\Models\User.
     *
     * @property int $id
     * @property string $name
     * @property \Illuminate\Support\Carbon|null $created_at
     * @property \Illuminate\Support\Carbon|null $updated_at
     * @property \Illuminate\Support\Carbon|null $deleted_at
     * @property ?\App\Models\Team $currentTeam
     * @property-read \Illuminate\Database\Eloquent\Collection|\App\Models\Team[] $teams
     * @method static \Illuminate\Database\Query\Builder|User withTrashed()
     * @method static \Illuminate\Database\Query\Builder|User withoutTrashed()
     */
    class User extends \Eloquent
    {
    }
```

**Comment by szepeviktor** [comment]
Here you go Jeffrey!

`public function teams(): BelongsToMany`

**Reply by JeffreyDavidson** [reply]
After applying php 8 return type of BelongsToMany and rerunning Larasstan with `./vendor/bin/phpstan analyse --memory-limit=2G app` I'm still getting the same error from Larastan.

**Comment by szepeviktor** [comment]
Okay.
What if you try defining User this way https://github.com/laravel/laravel/blob/74dfb6cec4d2c1f6d7e1a1e29e4bc9d610dc7031/app/Models/User.php#L3-L11 and drop all `@method` and `@property` tags?

**Reply by szepeviktor** [reply]
Actually what is your last code block?

**Reply by JeffreyDavidson** [reply]
My apologies its actually from ide_helper_models.php. The model I'm actually questioning isn't the exact User model. Its actually a different model. I just called it User to make it simpler to try and figure out.

**Reply by szepeviktor** [reply]
I see!
Please disable ide-helper and remove all of its traces.
Then Larastan will work.

**Reply by JeffreyDavidson** [reply]
@szepeviktor I've removed the ide-helper package and the file and reran it and Larastan is still reporting the error of accessing an undefined currentTeam property.

**Comment by szepeviktor** [comment]
What if you temporarily move `function currentTeam()` accessor from the trait to the model?

**Reply by JeffreyDavidson** [reply]
Did so but for some reason still getting the same error. Seems odd.

**Reply by szepeviktor** [reply]
The last chance is to remove packages and source code step-by-step from your project and see when this error goes away.
Remember to clear PHPStan' s cache: `phpstan clear-result-cache`

**Reply by szepeviktor** [reply]
No! Please see this
https://github.com/nunomaduro/larastan/blob/fa598dbc2b319673857ee17f6afe1aed6564b505/tests/Application/app/User.php#L163-L191
- it has to be protected
- and use PHP generics

**Comment by canvural** [comment]
Hi,

To be able to recognize Laravel 9 attributes, you need to make your attribute method `protected` And also add generic return type to the PHPDoc block. For example `/** @return Attribute<string[], string[]> */`

**Reply by tusharnain** [reply]
Thank you very much!
